src/oph/ehoks/palaute/opiskelija.clj

```diff
@@ -128,22 +104,43 @@
   ([{:keys [hoks opiskeluoikeus] :as ctx} kysely-type opts]
     (jdbc/with-db-transaction
       [tx db/spec]
-      (let [ctx     (assoc ctx
-                           :tx              tx
-                           :koulutustoimija (palaute/koulutustoimija-oid!
+      (let [ctx (assoc ctx :koulutustoimija (palaute/koulutustoimija-oid!
                                               opiskeluoikeus))
-            palaute {:type kysely-type
-                     :existing-heratteet
-                     (when-not (:resend? opts)
-                       (existing-heratteet! ctx kysely-type))}
-            [state field reason] (initial-palaute-state-and-reason ctx palaute)
-            tapahtuma {:reason     reason
-                       :other-info (select-keys hoks [field])}]
+            existing-heratteet (when-not (:resend? opts)
+                                 (existing-heratteet! tx ctx kysely-type))
+            [state field reason] (initial-palaute-state-and-reason
+                                   ctx kysely-type existing-heratteet)]
         (log/info "Initial state for" kysely-type "for HOKS" (:id hoks)
                   "will be" (or state :ei-luoda-ollenkaan)
                   "because of" reason "in" field)
         (when state
-          (initiate! ctx (assoc palaute :state state :tapahtuma tapahtuma)))
+          (let [heratepvm       (get hoks (herate-date-basis kysely-type))
+                existing-herate (find-first palaute/unhandled?
+                                            existing-heratteet)
+                target-kasittelytila (not= state :odottaa-kasittelya)
+                amisherate-kasittelytila
+                (db-hoks/get-or-create-amisherate-kasittelytila-by-hoks-id!
+                  (:id hoks))]
+            (db-hoks/update-amisherate-kasittelytilat!
+              tx {:id (:id amisherate-kasittelytila)
+                  (kysely-kasittely-field-mapping kysely-type)
+                  target-kasittelytila})
+            (->> (build! ctx kysely-type state existing-herate)
+                 (palaute/upsert! tx)
+                 (palautetapahtuma/build ctx state field reason existing-herate)
+                 (palautetapahtuma/insert! tx))
```

Comment:
tää faktorointi on sinänsä hyvä mutta mieluusti kyllä edelleen nostaisin myös tän koodinpätkän erilleen `initiate-if-needed!`-funktiosta ihan jo sitäkin varten että se saa nimen (kuten sync-palaute-to-db! tjsp)

Reply:
vakavuus: pyyntö

Reply:
Vastaan tällä kommentilla tähän sekä tuohon sun ylempänä antamaan katselmointikommenttiin koskien `initiate-if-needed?`-funktion pituutta.

Itse oon jotenkin aina nähnyt, että funktioiden saaminen lyhyemmäksi ei oo sinänsä itseisarvo, mutta koodin selkeyttä ja helppolukuisuutta oon pitäny tärkeänä. Nähdäkseni funktioilla ensisijaisesti vältetään toisteista koodia, ja toissijaisesti abstraktoidaan isompia koodiblokkeja, _sillä edellytyksellä että funktio pystytään nimeämään yksiselitteisesti_, koska muuten prosessissa tulee vaan siirtäneeksi koodia paikasta X paikkaan Y, ja oman kokemukseni mukaan koodikokonaisuudesta tulee näin hivenen hankalampi lukea (ympäripyöreästi nimettyjen funkkareiden tapauksessa joutuu tekemään paljon editorissa go-to-definitionia, ja itsellä ainakin aivoissa tapahtuu tässä kohtaa jonkinmoinen kontekstin vaihdos :sweat_smile:) Itse oon näin ollen toiminu periaatteella, että jos funktiota ei mitenkään yksiselitteisesti pysty nimeämään, niin koodin eriyttäminen funktioks on todennäköisesti parempi jättää tekemättä, ellei tällä sitten just vältetä toisteista koodia. Ts. koen siis, että on helpompi lukea pitkää funktiota rivi riviltä, kuin moneen pienempään osaan pilkottua funktiota, jossa ne osat on huonosti abstraktoituja.

Tuosta `sync-palaute-to-db!`-nimestä jää mun nähdäkseni uupumaan se, että palautteen lisäksi kantaan tallennetaan myös palautetapahtuma, jolloin nimeksi olisi hyvä vähintäänkin laittaa `sync-palaute-and-palautetapahtuma-to-db!`.  Mutta toisaalta tuostakaan funktionimestä ei käy ilmi, millä menetelmin (insert, update, upsert?) palaute ja palautetapahtuma "synkataan", tässä tapauksessa palaute upsertilla ja palautetapatuma insertilla, joten nimi vois olla myöskin `upsert-palaute-and-insert-palautetapahtuma-to-db!`. Tälläisissa tapauksissa, kun itse funktion nimellä alkaa olla mittaa paljon suhteessa koodiblokkiin jonka se abstraktoi, tuoden muutenkin melko vähän lisäarvoa, oon usein jättäny koodin eriyttämättä omaksi funktiokseen.

Reply:
> sillä edellytyksellä että funktio pystytään nimeämään yksiselitteisesti, koska muuten prosessissa tulee vaan siirtäneeksi koodia paikasta X paikkaan Y,

Jep, tää on hyvä huomio.  Vielä yksi ehdotus nimeksi: `save-palaute-state!`